### PR TITLE
Prob trace extension

### PR DESCRIPTION
**Pull request by tgymnich**
- add `insertArgument`
- add `insertReturn`
- add  `insertFunction(Pointer)`

### CODE REVIEW

**Comment by tgymnich** [comment]
fixed in #1065

File: enzyme/Enzyme/TraceInterface.cpp [hunk 1]
```diff
@@ -110,6 +133,8 @@ FunctionType *TraceInterface::hasChoiceTy(LLVMContext &C) {
 StaticTraceInterface::StaticTraceInterface(Module *M)
     : TraceInterface(M->getContext()) {
   for (auto &&F : M->functions()) {
+    if (F.isIntrinsic())
```
Comment (wsmoses) [inline enzyme/Enzyme/TraceInterface.cpp:136]: Duplicate?

Comment (tgymnich) [reply enzyme/Enzyme/TraceInterface.cpp:136]: oops

File: enzyme/Enzyme/TraceInterface.cpp [hunk 2]
```diff
@@ -110,6 +133,8 @@ FunctionType *TraceInterface::hasChoiceTy(LLVMContext &C) {
...
+      continue;
```
Comment (tgymnich) [inline enzyme/Enzyme/TraceInterface.cpp:137]:
```suggestion
```

File: enzyme/Enzyme/Utils.h [hunk 3]
```diff
@@ -1044,6 +1046,8 @@ template <typename T> static inline llvm::StringRef getFuncNameFromCall(T *op) {
       return called->getFnAttribute("enzyme_math").getValueAsString();
     else if (called->hasFnAttribute("enzyme_allocator"))
       return "enzyme_allocator";
+    else if (called->hasFnAttribute("enzyme_notypeanalysis"))
```
Comment (wsmoses) [inline enzyme/Enzyme/Utils.h:1049]: This should not be here, this is for use in adjointgenerator/etc. You should modify TypeAnalysis to do this check.

File: enzyme/Enzyme/TraceGenerator.cpp [hunk 4]
```diff
@@ -109,6 +119,11 @@ void TraceGenerator::handleSampleCall(CallInst &call, CallInst *new_call) {
     new_call->replaceAllUsesWith(outline_call);
     new_call->eraseFromParent();
 
+    auto density_fn = ValueAsMetadata::get(likelihoodfn);
+    auto density_node = MDNode::get(outline_call->getContext(), {density_fn});
+
+    outline_call->setMetadata("pdf", density_node);
```
Comment (wsmoses) [inline enzyme/Enzyme/TraceGenerator.cpp:125]: prefix to enzyme_pdf ?

**Review by wsmoses — APPROVED** [review]
It would be nice to also add a test to check the metadata inserted being conforming

File: enzyme/Enzyme/TraceGenerator.cpp (hunk 4 above)
Comment (tgymnich) [reply enzyme/Enzyme/TraceGenerator.cpp]: done

**Comment by tgymnich** [comment]
> It would be nice to also add a test to check the metadata inserted being conforming

@wsmoses I would consider the metadata an implementation detail for now. It thus does not require testing in my opinion.

**Pull request merged**